### PR TITLE
Hyprlock: add module

### PR DESCRIPTION
**Pull request by khaneliman**
### Description
Adding the hyprlock module, similar to https://github.com/nix-community/home-manager/pull/5324

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC
@fufexan

### CODE REVIEW

**Comment by khaneliman** [comment]
@rycee rebased and resolved merge conflicts

File: modules/programs/hyprlock.nix [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            }
+          ];
+        }
+      '';
+    };
+
+    extraConfig = lib.mkOption {
+      type = lib.types.lines;
+      default = "";
+      example = ''
```
Comment (andresilva) [inline modules/programs/hyprlock.nix:91]: This example is unrelated to hyprlock

File: modules/programs/hyprlock.nix [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    sourceFirst = lib.mkEnableOption ''
+      putting source entries at the top of the configuration
+    '' // {
+      default = true;
+    };
+
+    importantPrefixes = lib.mkOption {
+      type = with lib.types; listOf str;
+      default = [ "$" "monitor" "size" ]
+        ++ lib.optionals cfg.sourceFirst [ "source" ];
```
Comment (andresilva) [inline modules/programs/hyprlock.nix:109]: Can you actually source files in hyprlock config? (not sure if this only applies to hyprland)

**Review by andresilva — CHANGES_REQUESTED** [review]

File: modules/programs/hyprlock.nix (hunk 2 above)
Comment (khaneliman) [reply modules/programs/hyprlock.nix:109]: https://github.com/hyprwm/hyprlock/blob/c87af3aa1f6e6bd06cffaabcc400bd45e26d565a/nix/hm-module.nix#L74

Comment (andresilva) [reply modules/programs/hyprlock.nix:109]: Yep, I checked the hyprlock code, it's supported.

File: modules/programs/hyprlock.nix (hunk 1 above)
Comment (andresilva) [reply modules/programs/hyprlock.nix:91]: `extraConfig` is needed for Hyprland since you can't define submaps in the generic settings attribute. For hyprlock not sure if it's needed, but might be useful as an escape hatch to just define stuff plainly.

File: modules/programs/hyprlock.nix [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        security.pam.services.hyprlock = {};
+        ```
+      '';
+    };
+
+    package = mkOption {
+      description = "The hyprlock package";
+      type = types.package;
+      default = pkgs.hyprlock;
+    };
```
Comment (sumnerevans) [inline modules/programs/hyprlock.nix:29]:
```suggestion
    package = mkPackageOption pkgs "hyprlock" { };
```

**Review by andresilva — APPROVED** [review]

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Comment by nyabinary** [comment]
https://github.com/hyprwm/hyprlock/blob/main/nix/hm-module.nix
Does this reach feature parity with this?

**Comment by khaneliman** [comment]
> https://github.com/hyprwm/hyprlock/blob/main/nix/hm-module.nix Does this reach feature parity with this?

It is done slightly differently, that had nix named attributes and validation unique to each setting. This just validates you're passing in valid nix that can be converted into hyprlang. You can see a linked commit above where I migrated from the flake hm-module to using the home-manager module. Main difference being everything is nested under a settings attribute, the names for attributes need to match the actual hyprlock config names, and there are no default values injected.
